src/lib.rs

```diff
@@ -82,7 +82,7 @@ mod tests {
     }
 
     #[test]
-    fn two_async_elements_no_waiting() {
+    fn two_async_elements_immediate_notify() {
```

Comment:
I see the point of changing the test name to signify the change in polling behavior, but that isn't what the `no_waiting` was originally specifying. `no_waiting` was referring to the packet generator input immediately yielding values (i.e. never yielding `Async::NotReady`) until it was empty (then sends `Async::Ready(None)`). To verify that the new `task::notify` behavior works correctly, let's construct a new test called `<element_arrangement>_source_packets_generated_on_interval` that uses a `LinearPacketGenerator` rather than `iter_ok`. 

It would be even nicer if we could specify `context`s like in RSpec so we could say

```
context "packets arrive every #{source_interval} seconds" do
  let packet_generator = LinearIntervalGenerator::new(time::Duration::from_millis(source_interval), 10);

  // ... all test examples that use the packet generator
end
```

I'm not sure this is possible with Rust's test framework though. If contexts don't exist, we'll just have to use long test names, which are only slightly more inconvenient.

Reply:
Okay, I'll put in another commit to add the Linear Generator test, fixup the naming convention. Standby

Reply:
I had to push this as a new commit, since we remove the ForeverDrain in the 2nd commit and I did not want to fix the merge conflicts of squashing 1 and 3 together.  I'm not sure about making the tests have a bit more Rspec way, maybe someone else could look into that. I'm sure that the very least we could define common functions.